# Select key Biomarkers

## 1. Import necessary libraries, have function definitions

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
from sklearn.utils import resample

import warnings
warnings.filterwarnings("ignore")

In [2]:
# Train a LASSO logistic regression model with cross-validation and stability selection
def train_lasso_model(X_train, y_train, num_bootstraps=50, selection_threshold=0.5):

    # Standardize the training data
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)

    # Define a LASSO model (LogisticRegression with L1 penalty)
    lasso = LogisticRegression(penalty="l1", solver="liblinear", max_iter=1000)

    # Grid search for the best regularization strength
    param_grid = {"C": np.logspace(-8, -1, 40)}
    grid_search = GridSearchCV(lasso, param_grid, cv=5, scoring="accuracy", n_jobs=-1)
    grid_search.fit(X_train_scaled, y_train)
    best_C = grid_search.best_params_["C"]
    best_model = LogisticRegression(penalty="l1", solver="liblinear", C=best_C, max_iter=1000)
    
    # Perform Stability Selection (Bootstrap)
    selection_counts = np.zeros(X_train.shape[1])
    for _ in range(num_bootstraps):
        X_resampled, y_resampled = resample(X_train_scaled, y_train, random_state=_)
        best_model.fit(X_resampled, y_resampled)
        selection_counts += (best_model.coef_[0] != 0)

    # Compute selection frequency
    selection_freq = selection_counts / num_bootstraps

    # Select biomarkers that appear in at least `selection_threshold` % of bootstrap runs
    selected_biomarkers = X_train.columns[selection_freq >= selection_threshold]

    # Store LASSO results
    lasso_results = pd.DataFrame({
        "Biomarker": X_train.columns,
        "Coefficient": best_model.coef_[0],
        "Selection_Frequency": selection_freq
    })

    return best_model, scaler, selected_biomarkers, lasso_results


In [3]:
# Evaluate the trained LASSO model on the validation set using ROC AUC.
def evaluate_lasso_model(model, scaler, X_val, y_val):
    X_val_scaled = scaler.transform(X_val)
    y_val_pred_prob = model.predict_proba(X_val_scaled)
    
    # If binary classification , use the probability of the positive class
    if y_val_pred_prob.shape[1] == 2:
        auc_val = roc_auc_score(y_val, y_val_pred_prob[:, 1])
    else:
        auc_val = roc_auc_score(y_val, y_val_pred_prob, multi_class="ovr")
    return auc_val

## 2. Load the data

In [4]:
# Load the biomarker data that we cleaned and scaled in N1_biomarker_data_cleaning.py
biomarker_df_file_path = "../data/clean_data/scaled_biomarker_data_with_subjectid.csv"
biomarker_df = pd.read_csv(biomarker_df_file_path)
biomarker_df["subject_id"] = biomarker_df["subject_id"].astype(str)

print(biomarker_df.shape)
biomarker_df.head()

(466, 21)


subject_id  ...  balf_TNF-RI_V1_imputed
0       3901  ...               -1.412866
1       3695  ...                0.753702
2       4097  ...                0.560863
3       3738  ...               -1.044524
4       3791  ...                0.599358

[5 rows x 21 columns]

In [5]:
# Load the clusters from LCA analysis
file_path = "../data/clean_data/vap_cluster_assignments_k4_lca.csv"
df_cluster = pd.read_csv(file_path)
df_cluster["subject_id"] = df_cluster["subject_id"].astype(str)
df_cluster.drop(columns=["Unnamed: 0"], inplace=True)

print(df_cluster.shape)
df_cluster.head()

(466, 2)


subject_id  cluster
0       3901        3
1       3695        4
2       4097        4
3       3738        3
4       3791        4

In [6]:
# Merge the biomarker data with the cluster assignments
biomarker_df_with_clusters = biomarker_df.merge(df_cluster, on="subject_id", how="inner")
print(biomarker_df_with_clusters.shape)
biomarker_df_with_clusters

(465, 22)


subject_id  balf_Amphiregulin_V1_imputed  ...  balf_TNF-RI_V1_imputed  cluster
0         3901                     -1.470130  ...               -1.412866        3
1         3695                      0.803770  ...                0.753702        4
2         4097                      0.553875  ...                0.560863        4
3         3738                     -1.671626  ...               -1.044524        3
4         3791                      1.005517  ...                0.599358        4
..         ...                           ...  ...                     ...      ...
460       3564                      0.136118  ...                0.345492        1
461       3884                     -0.572252  ...               -0.764395        1
462       4057                     -1.358552  ...               -1.466420        3
463       3682                     -1.160260  ...               -0.526094        1
464       3765                      0.076017  ...               -0.631772        1

[465 rows x 22 columns]

## 4. LASSO modeling to select the key biomarkers

Next, let us see which are the key biomarkers.


In [7]:
X = biomarker_df_with_clusters.drop(columns=["cluster", "subject_id"])
y = biomarker_df_with_clusters["cluster"]

# Perform a 80-20 train val split
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=30)

# Train the LASSO model and extract selected biomarkers
best_model, scaler, selected_biomarkers, lasso_results = train_lasso_model(X_train, y_train)

print("\nSelected ", len(selected_biomarkers), " Biomarkers:")
print(selected_biomarkers)

print("\nLASSO Results:")
print(lasso_results)



Selected  5  Biomarkers:
Index(['balf_IP-10_chemo_V1_imputed', 'balf_TARC_chemo_V1_imputed',
       'balf_GM-CSF_V1_imputed', 'balf_IL-15_V1_imputed',
       'balf_TNF-RI_V1_imputed'],
      dtype='object')

LASSO Results:
                         Biomarker  Coefficient  Selection_Frequency
0     balf_Amphiregulin_V1_imputed     0.000000                 0.08
1     balf_Calprotectin_V1_imputed     0.000000                 0.20
2            balf_CD163_V1_imputed     0.000000                 0.00
3       balf_IL-8_chemo_V1_imputed     0.000000                 0.34
4      balf_IP-10_chemo_V1_imputed     0.000000                 0.76
5      balf_MCP-1_chemo_V1_imputed     0.076553                 0.42
6      balf_MCP-4_chemo_V1_imputed     0.090808                 0.40
7       balf_TARC_chemo_V1_imputed     0.010921                 0.50
8           balf_GM-CSF_V1_imputed     0.000000                 0.50
9   balf_IL-12/IL-23p40_V1_imputed     0.000000                 0.48
10           balf

The LASSO logistic regression with stability selection identified five key biomarkers, with `balf_TNF-RI_V1_imputed` (100%) as the most stable predictor, showing a strong inverse relationship with mortality (coefficient = -0.43). Other frequently selected biomarkers included `balf_IP-10_chemo_V1_imputed` (76%), `balf_TARC_chemo_V1_imputed` (50%), `balf_GM-CSF_V1_imputed` (50%), and `balf_IL-15_V1_imputed` (60%), suggesting their potential clinical relevance.

Meanwhile, biomarkers like `balf_Amphiregulin_V1_imputed` and `balf_CD163_V1_imputed` were rarely or never selected, indicating limited predictive value.

In [8]:
# Evaluate the model
auc_val = evaluate_lasso_model(best_model, scaler, X_val, y_val)
print("\nValidation AUC:", auc_val)


Validation AUC: 0.9576380544206787


With a Validation AUC of 0.96, our model seems to be performing really well in distinguishing outcomes. We’ll now dive deeper into statistical analysis in the coming notebooks to understand how these biomarkers and clusters are linked to hospital mortality and other key clinical factors.